src/nbl/asset/utils/IShaderCompiler.cpp

```diff
@@ -289,3 +219,116 @@ auto IShaderCompiler::CIncludeFinder::tryIncludeGenerators(const std::string& in
 
     return {};
 }
+
+core::smart_refctd_ptr<asset::ICPUShader> IShaderCompiler::CCache::find(const SEntry& mainFile, const IShaderCompiler::CIncludeFinder* finder) const
+{
+    auto foundRange = m_container.equal_range(mainFile);
+    for (auto& found = foundRange.first; found != foundRange.second; found++)
+    {
+        bool allDependenciesMatch = true;
+        // go through all dependencies
+        for (auto i = 0; i < found->dependencies.size(); i++)
+        {
+            const auto& dependency = found->dependencies[i];
+
+            IIncludeLoader::found_t header;
+            if (dependency.standardInclude)
+                header = finder->getIncludeStandard(dependency.requestingSourceDir, dependency.identifier);
+            else
+                header = finder->getIncludeRelative(dependency.requestingSourceDir, dependency.identifier);
+
+            if (header.hash != dependency.hash || header.contents != dependency.contents)
+            {
+                allDependenciesMatch = false;
+                break;
+            }
+        }
+        if (allDependenciesMatch) {
+            return found->value;
+        }
+    }
+    return nullptr;
+}
+
+core::smart_refctd_ptr<ICPUBuffer> IShaderCompiler::CCache::serialize() const
+{
+    core::vector<uint8_t> shadersBuffer;
+    json entries;
+    json shaderCreationParams;
+    for (auto& entry : m_container) {
+        // Add the entry as a json array
+        entries.push_back(entry);
+
+        // Now create the CPU shader creation parameters struct
+        CPUShaderCreationParams params;
+        params.stage = entry.value->getStage();
+        params.contentType = entry.value->getContentType();
+        params.filepathHint = entry.value->getFilepathHint();
+        params.offset = shadersBuffer.size();
+        params.codeByteSize = entry.value->getContent()->getSize();
+
+        // And add it to the shader creation parameters array
+        shaderCreationParams.push_back(params);
+
+        // Finally, insert the shader bytecode into the shaders buffer
+        shadersBuffer.reserve(shadersBuffer.size() + params.codeByteSize);
+        shadersBuffer.insert(shadersBuffer.end(), (uint8_t*)entry.value->getContent()->getPointer(), (uint8_t*)entry.value->getContent()->getPointer() + params.codeByteSize);
+    }
+    json containerJson{
+        { "entries", entries },
+        { "shaderCreationParams", shaderCreationParams },
+    };
+    std::string dumpedContainerJson = std::move(containerJson.dump());
+    uint64_t dumpedContainerJsonLength = dumpedContainerJson.size();
+    core::vector<uint8_t> retVal;
```

Comment:
if you saved the shader data size and rawarray first, you could only have one `vector` and two `resize` calls